### ISSUE TITLE
fatalnvml: Insufficient Permissions

### ISSUE THREAD

**Issue by qingshanyinyin** [NVIDIA/k8s-device-plugin#201]
Firstly, I set the mig of GPUs of  **dev0** （**7*5s**）:
![image](https://user-images.githubusercontent.com/35218255/96715897-d6a50480-13d6-11eb-8d4e-4debc9ed240f.png)

when it call the func (**nvml.NewDevice(i)**):
![image](https://user-images.githubusercontent.com/35218255/96716032-12d86500-13d7-11eb-8a7b-09a9ccc6b01c.png)
NewDevice() (**k8s-device-plugin/vendor/github.com/NVIDIA/gpu-monitoring-tools\bindings\go\nvml\nvml.go**)
![image](https://user-images.githubusercontent.com/35218255/96716355-824e5480-13d7-11eb-8825-0819faf9f757.png)
newDevice() (**k8s-device-plugin/vendor/github.com/NVIDIA/gpu-monitoring-tools\bindings\go\nvml\nvml.go**)
![image](https://user-images.githubusercontent.com/35218255/96716398-8f6b4380-13d7-11eb-98e6-2fecb7b400a7.png)

The error is "**fatalnvml: Insufficient Permissions**"
![image](https://user-images.githubusercontent.com/35218255/96716290-6e0a5780-13d7-11eb-84cf-56458137ff7f.png)

I wonder to know  whether a card  can also use the func(nvml.NewDevice(i))  when it  is set to vGPUs？As k8s-device-plugin do:
![image](https://user-images.githubusercontent.com/35218255/96717412-fa694a00-13d8-11eb-8cf7-a37949e29e71.png)

Thanks!

**Comment by klueska** [comment]
The plugin never calls `nvml.NewDevice()` anywhere -- it only calls `nvml.NewDeviceLite()`, which shouldn't have this issue.

That said, the issue comes from the fact that you can't call `GetMemoryInfo()` without elevated permissions when running in MIG mode (which `nvml.NewDevice()` calls under the hood).

Therefore, when calling this function with MIG mode enabled, you need to (at a minimum) have the privileges granted to you by `mig-monitor` capability. Details on this capability (when asking for it using the nvidia-container-toolkit) can be found here: https://github.com/NVIDIA/nvidia-container-runtime#nvidia_mig_monitor_devices

Details on nvidia-capabilities in general can be found here:
https://docs.google.com/document/d/194A-Hg3mLlIW4eo2BSUcGKpzZf2ciX47eoH5T-WZNXo/edit#

**Comment by qingshanyinyin** [comment]
Thank u very much! I solve it as your URL!
I want to set the mig in the **k8s-device-plugin** pod of k8s! When i set the mig mod=1,the error is "MIG mod is in pending enanble state"! Because some processes in the card（**kubelet,****k8s-devcie-plugin**） ,as bellow:
![image](https://user-images.githubusercontent.com/35218255/96845256-145f6700-1483-11eb-8e38-8029cab6260c.png)
![image](https://user-images.githubusercontent.com/35218255/96845500-58eb0280-1483-11eb-973e-52decadcb697.png)
Is there any way to config the vGPUs In the plugin Pod of k8s as your mig.go (gpu-monitoring-tools/bings/go/nvml/mig.go) if i donn't stop the kubelet! @klueska

**Comment by klueska** [comment]
No, you must stop the kubelet. You must also stop the device plugin (and any other clients of the GPU, e.g. gpu-feature-discovery, etc.).

Please see the discussion starting here:
https://github.com/NVIDIA/k8s-device-plugin/issues/180#issuecomment-682344586

**Comment by klueska** [comment]
Also, if what you are trying to do is run the plugin with MIG support, this is a good resource:
https://docs.google.com/document/d/1bshSIcWNYRZGfywgwRHa07C0qRyOYKxWYxClbeJM-WM/edit

**Comment by qingshanyinyin** [comment]
I have almost achieve the plan that automaticly setting vGPUs in device-plugin! But there is a problem which show device-plugin doesn't have the **suffient permissions** when call the function of **CreateComputeInstance()** ! And when the migMode is pending, there is no way (interface in mig.go) to solve it but cmd(nvidia-smi --gpu-reset)!

**Comment by klueska** [comment]
Just to be clear, `vGPU` is a very different technology than `MIG`. It seems you are trying to configure `MIG` here and not `vGPUs`.

In order to create GIs and CIs, you need to have proper access to the `mig-config` capability as described here:
https://docs.google.com/document/d/194A-Hg3mLlIW4eo2BSUcGKpzZf2ciX47eoH5T-WZNXo/edit#

You cannot reset a GPU via NVML because you need to drop your handle to NVML before doing so. You need to call directly into the driver (via an `ioctl()`) to perform the reset, and the `nvidia-smi` is a good place to do this from since it already wraps this call for you.

**Comment by qingshanyinyin** [comment]
If i shutdown the nvidia-fabricmanager service（using GPU process , "lsof /dev/nvidia*" can see）, and the mig mod of GPU is pendding as bellow:
![image](https://user-images.githubusercontent.com/35218255/100444688-ce7f6980-30e6-11eb-8c27-281153e089a7.png)
so i reset the gpus,and will be an error in the reset gpu (nvidia-smi --gpu-reset) ! But it will be setted successfully, as bellow:
![image](https://user-images.githubusercontent.com/35218255/100444142-ee625d80-30e5-11eb-8df1-e66ec5c02cc7.png)
If i donn‘t shutdown the nvidid-fabricmanager service, the time of setting mig is much longger than shutdown it ！
I wonder to know wether i should shutdown the nvidia-fabricmanager service when Set MIG?  When i open the nvidia-fabricmanager, calling the interface (such as SetMigMode()、CreateGPUInstance ...)  costs much time than close it! Why?Thanks!

**Comment by klueska** [comment]
I'm sorry, I don't quite understand your question. Can you rephrase it please?

**Comment by qingshanyinyin** [comment]
Thanks!
Sorry, I have not get enough pictures of the problem!  In one word, should i kill the process of nvidia-fabricmanager which use gpu (lsof /dev/nvidia* can find) when i set the mig of A100 GPU ？Does this process affect the speed of setting mig of A100 GPU?@klueska

**Comment by qingshanyinyin** [comment]
I have annother question. 
Firstly, i set the A100 GPUs (8 cards) to 8*[MIG-7g.40gb]:
![image](https://user-images.githubusercontent.com/35218255/100994185-4ccd8700-3591-11eb-8b94-e8d47e88c275.png)
Secondly, i choose two MIG-UUID to run tensorflow-cuda11 as https://docs.google.com/document/d/13LyeFxuWLSwDjCXKx7RyWWRZEAqA-mghtK3GRPvGv1s/edit:
![image](https://user-images.githubusercontent.com/35218255/100994637-c1082a80-3591-11eb-9293-65001001065b.png)

**Comment by klueska** [comment]
> should i kill the process of nvidia-fabricmanager which use gpu (lsof /dev/nvidia* can find) when i set the mig of A100 GPU

(Almost) all clients of the GPU driver must be killed before you can issue a gpu-reset. Some clients have been updated to perform a handshake with the driver so that they don't need to be killed, and a reset will succeed. I am not sure, if the `fabricmanager` is one of these or not. I would kill/restart it to be on the safe side.

> Does this process affect the speed of setting mig of A100 GPU?

It shouldn't.

**Comment by klueska** [comment]
> I have annother question.

I didn't see an explicit question, but I'm guessing you are confused why tensorflow isn't using both MIG devices you passed it? 

At present, all CUDA cleints (of which tensorflow is one) can only make use of a single MIG device at a time (because there is no P2P communication possible between MIG devices). It will simply pick up and use the first one it finds, ignoring all others.

If you actually want to inject multiple MIG devices into a container and make use of them, you will have to run separate CUDA clients (i.e. separate processes) that consume them independently. You can set the environment variable CUDA_VISIBLE_DEVICES to control which process has access to which device.

**Comment by qingshanyinyin** [comment]
> > I have annother question.
> 
> I didn't see an explicit question, but I'm guessing you are confused why tensorflow isn't using both MIG devices you passed it?
> 
> At present, all CUDA cleints (of which tensorflow is one) can only make use of a single MIG device at a time (because there is no P2P communication possible between MIG devices). It will simply pick up and use the first one it finds, ignoring all others.
> 
> If you actually want to inject multiple MIG devices into a container and make use of them, you will have to run separate CUDA clients (i.e. separate processes) that consume them independently. You can set the environment variable CUDA_VISIBLE_DEVICES to control which process has access to which device.

Thanks!  I am editting!

**Comment by qingshanyinyin** [comment]
There is no P2P communication possible between MIG devices! if they can communicate pass cpu between two MIG devices？@klueska
